python-checks/src/main/java/org/sonar/python/checks/MandatoryFunctionParameterTypeHintCheck.java

```diff
@@ -41,28 +44,41 @@ public void initialize(Context context) {
       FunctionDefImpl functionDef = (FunctionDefImpl) ctx.syntaxNode();
       ParameterList parameterList = functionDef.parameters();
       FunctionSymbol functionSymbol = functionDef.functionSymbol();
-      if (parameterList != null ) {
-        removeSelfAndClassParameters(parameterList, functionSymbol)
-          .forEach(parameter -> {
-            if (parameter.typeAnnotation() == null) {
-              ctx.addIssue(parameter, MESSAGE);
-            }
-          });
+      if (parameterList != null) {
+        List<Parameter> parameters = parameterList.nonTuple();
+        for (int i = 0; i < parameters.size(); i++) {
+          Parameter parameter = parameters.get(i);
+          boolean isFirstParameter = i == 0;
+          if (!isASelfInstanceParameter(parameter, functionSymbol) &&
+            !isFirstParamOfClassAnnotatedMethod(functionSymbol, isFirstParameter) &&
+            !isFirstParamOfNewMethod(functionSymbol, isFirstParameter) &&
+            !isSpecialCharParameter(parameter) && parameter.typeAnnotation() == null) {
+            ctx.addIssue(parameter, MESSAGE);
+          }
+        }
       }
     });
   }
 
-  private static Stream<Parameter> removeSelfAndClassParameters(ParameterList parameterList, @Nullable FunctionSymbol functionSymbol) {
-    return parameterList.nonTuple().stream()
-      .filter(param -> !isASelfInstanceParameter(param, functionSymbol) && !isAClassMethodParameter(param, functionSymbol));
-  }
-
   private static boolean isASelfInstanceParameter(Parameter parameter, @Nullable FunctionSymbol functionSymbol) {
     return hasName("self", parameter) && functionSymbol != null && functionSymbol.isInstanceMethod();
   }
 
-  private static boolean isAClassMethodParameter(Parameter parameter, @Nullable FunctionSymbol functionSymbol) {
-    return hasName("cls", parameter) && functionSymbol != null && functionSymbol.decorators().contains("classmethod");
+  private static boolean isFirstParamOfClassAnnotatedMethod(@Nullable FunctionSymbol functionSymbol, boolean isFirstParameter) {
+    return functionSymbol != null && functionSymbol.decorators().contains("classmethod") && isFirstParameter;
+  }
+
+  private static boolean isFirstParamOfNewMethod(@Nullable FunctionSymbol functionSymbol, boolean isFirstParameter) {
+    return functionSymbol != null && "__new__".equals(functionSymbol.name()) && isFirstParameter;
+  }
+
+  private static final List<String> specialTokenParams = Arrays.asList("*", "/");
+
+  private static boolean isSpecialCharParameter(Parameter parameter) {
+    Name parameterName = parameter.name();
+    Token token = parameter.starToken();
+    return (hasName("_", parameter)) ||
+      (parameterName == null && token != null && specialTokenParams.contains(token.value()));
```

Comment:
If you want to increase coverage here, you can use the `Optional.ofNullable` trick and map to the value, but I think it's fine as it is as well.